executor/benchmark_test.go

```diff
@@ -0,0 +1,344 @@
+	ctx    sessionctx.Context
+}
+
+type mockDataSource struct {
+	baseExecutor
+	p        mockDataSourceParameters
+	colData  [][]interface{}
+	chunks   []*chunk.Chunk
+	chunkPtr int
+}
+
+func (mds *mockDataSource) genColDatums(ordinal int) (results []interface{}) {
+	typ := mds.retFieldTypes[ordinal]
+	order := mds.p.orders[ordinal]
+	rows := mds.p.rows
+	NDV := mds.p.ndvs[ordinal]
+	results = make([]interface{}, 0, rows)
+	if NDV == 0 {
+		for i := 0; i < rows; i++ {
+			results = append(results, mds.randDatum(typ))
+		}
+	} else {
+		datumSet := make(map[string]bool, NDV)
+		datums := make([]interface{}, 0, NDV)
+		for len(datums) < NDV {
+			d := mds.randDatum(typ)
+			str := fmt.Sprintf("%v", d)
+			if datumSet[str] {
+				continue
+			}
+			datumSet[str] = true
+			datums = append(datums, d)
+		}
+
+		for i := 0; i < rows; i++ {
+			results = append(results, datums[rand.Intn(NDV)])
+		}
+	}
+
+	if order {
+		sort.Slice(results, func(i, j int) bool {
+			switch typ.Tp {
+			case mysql.TypeLong, mysql.TypeLonglong:
+				return results[i].(int64) < results[j].(int64)
+			case mysql.TypeDouble:
+				return results[i].(float64) < results[j].(float64)
+			default:
+				panic("not implement")
+			}
+		})
+	}
+
+	return
+}
+
+func (mds *mockDataSource) randDatum(typ *types.FieldType) interface{} {
+	switch typ.Tp {
+	case mysql.TypeLong, mysql.TypeLonglong:
+		return int64(rand.Int())
+	case mysql.TypeDouble:
+		return rand.Float64()
+	default:
+		panic("not implement")
+	}
+}
+
+func (mds *mockDataSource) prepareChunks() {
+	mds.chunks = make([]*chunk.Chunk, (mds.p.rows+mds.initCap-1)/mds.initCap)
+	for i := range mds.chunks {
+		mds.chunks[i] = chunk.NewChunkWithCapacity(mds.retTypes(), mds.ctx.GetSessionVars().MaxChunkSize)
+	}
+
+	for i := 0; i < mds.p.rows; i++ {
+		idx := i / mds.initCap
+		types := mds.retTypes()
+		for colIdx := 0; colIdx < len(types); colIdx++ {
+			switch types[colIdx].Tp {
+			case mysql.TypeLong, mysql.TypeLonglong:
+				mds.chunks[idx].AppendInt64(colIdx, mds.colData[colIdx][i].(int64))
+			case mysql.TypeDouble:
+				mds.chunks[idx].AppendFloat64(colIdx, mds.colData[colIdx][i].(float64))
+			default:
+				panic("not implement")
+			}
+		}
+	}
+	mds.chunkPtr = 0
+}
+
+func (mds *mockDataSource) Next(ctx context.Context, req *chunk.RecordBatch) error {
+	if mds.chunkPtr >= len(mds.chunks) {
+		req.Reset()
+		return nil
+	}
+	dataChk := mds.chunks[mds.chunkPtr]
+	dataChk.SwapColumns(req.Chunk)
+	mds.chunkPtr++
+	return nil
+}
+
+func buildMockDataSource(opt mockDataSourceParameters) *mockDataSource {
+	baseExec := newBaseExecutor(opt.ctx, opt.schema, "")
+	m := &mockDataSource{baseExec, opt, nil, nil, 0}
+	types := m.retTypes()
+	m.colData = make([][]interface{}, len(types))
+	for i := 0; i < len(types); i++ {
+		m.colData[i] = m.genColDatums(i)
+	}
+	return m
+}
+
+type aggTestCase struct {
+	// The test table's schema is fixed (aggCol Double, groupBy LongLong).
+	exec        string // "hash" or "stream"
```

Comment:
What about something like `execType` or `execStrategy` to make it more clear.

Reply:
fixed